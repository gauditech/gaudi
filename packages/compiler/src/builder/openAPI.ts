import _, { flatMap, mapValues } from "lodash";
import { OpenAPIV3 } from "openapi-types";
import { P, match } from "ts-pattern";

import { buildEndpointPath } from "@compiler/builder/query";
import { getRef } from "@compiler/common/refs";
import { FieldType } from "@compiler/compiler/ast/type";
import {
  endpointHasContext,
  endpointUsesAuthentication,
  endpointUsesAuthorization,
  getEndpointFieldset,
} from "@compiler/composer/entrypoints";
import {
  Definition,
  EndpointDef,
  EntrypointDef,
  FieldsetDef,
  SelectItem,
} from "@compiler/types/definition";

export function buildOpenAPI(definition: Definition): OpenAPIV3.Document {
  /**
   * Builds a list of required prop names that is attached to the object schema.
   *
   * Property is required if not nullable. This follows current validation implementation.
   */
  function buildRequiredProperties(select: SelectItem[]) {
    return select
      .map((select) => {
        return (
          match<typeof select, [string, boolean]>(select)
            .with({ kind: "expression" }, (item) => [item.alias, !item.type.nullable])
            // FIXME hook is currently always nullable
            .with({ kind: "model-hook" }, (item) => [item.alias, false])
            // reference/releation/query are always required themselves
            // their properties are handled for each object separately
            // FIXME respect cardinality and nullable!!
            .with({ kind: "nested-select" }, (item) => [item.alias, true])
            .exhaustive()
        );
      })
      .filter(([_name, required]) => required)
      .map(([name]) => name);
  }

  /** Create OpenAPI schema properties. */
  function buildSchemaProperties(select: SelectItem[]) {
    const schemaEntries = select.map((select): [string, OpenAPIV3.SchemaObject] => {
      return (
        match<typeof select, [string, OpenAPIV3.SchemaObject]>(select)
          .with({ kind: "expression" }, (item) => [
            item.alias,
            { type: convertToOpenAPIType(item.type.kind), nullable: item.type.nullable },
          ])
          // FIXME missing type info for hooks
          .with({ kind: "model-hook" }, (item) => [item.alias, {}])
          .with({ kind: "nested-select" }, (item) => {
            const ref = getRef(definition, item.refKey);
            // FIXME proper cardinality checks!
            const isObject = ref.kind === "reference";
            const properties = buildSchemaProperties(item.select);
            const required = buildRequiredProperties(item.select);
            if (isObject) {
              return [item.alias, { type: "object", properties, required }];
            } else {
              return [
                item.alias,
                { type: "array", items: { type: "object", properties, required } },
              ];
            }
          })
          .exhaustive()
      );
    });

    return Object.fromEntries(schemaEntries);
  }

  function buildEndpointOperation(
    endpoint: EndpointDef,
    apiName: string | undefined,
    parameters: OpenAPIV3.ParameterObject[]
  ): OpenAPIV3.OperationObject {
    const properties = buildSchemaProperties(endpoint.response ?? []);
    const required = buildRequiredProperties(endpoint.response ?? []);

    const objectSchema: OpenAPIV3.SchemaObject = { type: "object", properties, required };

    let responseSchema: OpenAPIV3.SchemaObject;
    if (endpoint.kind === "list") {
      // pageable list response
      if (endpoint.pageable) {
        responseSchema = {
          type: "object",
          properties: {
            page: { type: convertToOpenAPIType("integer") },
            pageSize: { type: convertToOpenAPIType("integer") },
            totalPages: { type: convertToOpenAPIType("integer") },
            totalCount: { type: convertToOpenAPIType("integer") },
            data: { type: "array", items: objectSchema },
          },
          required: ["page", "pageSize", "totalPages", "totalCount", "data"],
        };
      }
      // plain list response
      else {
        responseSchema = {
          type: "array",
          items: objectSchema,
        };
      }
    }
    // object response
    else {
      responseSchema = objectSchema;
    }

    const operation: OpenAPIV3.OperationObject = {
      tags: apiName ? [apiName] : undefined,
      responses: {},
    };

    if (endpoint.kind === "delete") {
      operation.responses[204] = {
        description: "Successful response",
      };
    } else {
      operation.responses[200] = {
        description: "Successful response",
        content: { "application/json": { schema: responseSchema } },
      };
    }

    if (endpointHasContext(endpoint)) {
      operation.responses[404] = {
        description: "Resource not found",
        content: {
          "application/json": {
            schema: {
              type: "object",
              properties: { message: { type: "string", enum: ["ERROR_CODE_RESOURCE_NOT_FOUND"] } },
            },
          },
        },
      };
    }

    if (endpointUsesAuthentication(endpoint)) {
      operation.responses[401] = {
        description: "Unauthenticated",
        content: {
          "application/json": {
            schema: {
              type: "object",
              properties: { message: { type: "string", enum: ["ERROR_CODE_UNAUTHENTICATED"] } },
            },
          },
        },
      };
    }

    if (endpointUsesAuthorization(endpoint)) {
      operation.responses[403] = {
        description: "Unauthorized",
        content: {
          "application/json": {
            schema: {
              type: "object",
              properties: { message: { type: "string", enum: ["ERROR_CODE_FORBIDDEN"] } },
            },
          },
        },
      };
    }

    operation.parameters = parameters;
    const fieldset = getEndpointFieldset(endpoint);
    if (fieldset) {
      const schema = buildSchemaFromFieldset(fieldset);
      operation.requestBody = { content: { "application/json": { schema } } };

      operation.responses[400] = {
        description: "Validation error",
        content: {
          "application/json": {
            schema: {
              type: "object",
              properties: {
                message: { type: "string", enum: ["Validation error"] },
                code: { type: "string", enum: ["ERROR_CODE_VALIDATION"] },
                // FIXME post validation v2:
                // * calculate `anyOf` for all the validation messages
                data: { type: "object" },
              },
            },
          },
        },
      };
    }

    if (endpoint.authorize) {
      operation.security = [{ bearerAuth: [] }];
    }

    return operation;
  }

  const paths: OpenAPIV3.PathsObject = {};
  definition.apis.forEach((api) =>
    api.entrypoints.forEach((entrypoint) =>
      extractEndpoints(entrypoint).forEach((endpoint) => {
        const endpointPath = buildEndpointPath(endpoint);
        const method = buildEndpointHttpMethod(endpoint);

        const path = [
          api.path,
          ..._.chain(endpointPath.fragments)
            .map((frag) => {
              return match(frag)
                .with({ kind: "namespace" }, (f) => f.name)
                .with({ kind: "identifier" }, (f) => `{${f.name}}`)
                .with({ kind: "query" }, () => null)
                .exhaustive();
            })
            .compact() // remove nulls
            .value(),
        ].join("/");

        const parameters = _.chain(endpointPath.fragments)
          .map((fragment): OpenAPIV3.ParameterObject | null => {
            return match(fragment)
              .with({ kind: "identifier" }, (f) => ({
                in: "path",
                name: f.name,
                required: true,
                schema: { type: convertToOpenAPIType(f.type) },
              }))
              .with({ kind: "query" }, (f) => ({
                in: "query",
                name: f.name,
                required: f.required,
                schema: { type: convertToOpenAPIType(f.type) },
              }))
              .with({ kind: "namespace" }, () => null)
              .exhaustive();
          })
          .compact()
          .value();

        const pathItem = paths[path] ?? {};
        pathItem[method] = buildEndpointOperation(endpoint, api.name, parameters);
        paths[path] = pathItem;
      })
    )
  );

  const securitySchemes: Record<string, OpenAPIV3.SecuritySchemeObject> = {};
  if (definition.authenticator) {
    securitySchemes["bearerAuth"] = { type: "http", scheme: "bearer" };
  }

  return {
    openapi: "3.0.3",
    info: { title: "Title", version: "1.0.0" },
    paths,
    components: {
      securitySchemes,
    },
  };
}

function buildSchemaFromFieldset(fieldset: FieldsetDef): OpenAPIV3.SchemaObject {
  if (fieldset.kind === "field") {
    return { type: convertToOpenAPIType(fieldset.type) };
  }

  const schema: OpenAPIV3.SchemaObject = { type: "object" };

  schema.properties = mapValues(fieldset.record, buildSchemaFromFieldset);

  const required = flatMap(fieldset.record, (fieldset, name) => (!fieldset.nullable ? [name] : []));
  if (required.length > 0) {
    schema.required = required;
  }

  return schema;
}

function convertToOpenAPIType(type: FieldType | "null"): OpenAPIV3.NonArraySchemaObjectType {
  switch (type) {
    case "boolean":
    case "integer":
    case "string":
      return type;
    case "float":
      return "number";
    case "null":
      // arbitrary type is "object"
      return "object";
  }
}

function extractEndpoints(entrypoint: EntrypointDef): EndpointDef[] {
  const nestedEndpoints = entrypoint.entrypoints.reduce((agg, entrypoint) => {
    return agg.concat(entrypoint.endpoints, extractEndpoints(entrypoint));
  }, [] as EndpointDef[]);
  return [...entrypoint.endpoints, ...nestedEndpoints];
}

function buildEndpointHttpMethod(endpoint: EndpointDef): OpenAPIV3.HttpMethods {
  switch (endpoint.kind) {
    case "list":
      return OpenAPIV3.HttpMethods.GET;
    case "get":
      return OpenAPIV3.HttpMethods.GET;
    case "create":
      return OpenAPIV3.HttpMethods.POST;
    case "update":
      return OpenAPIV3.HttpMethods.PATCH;
    case "delete":
      return OpenAPIV3.HttpMethods.DELETE;
    default: {
      const method = endpoint.method;
      return OpenAPIV3.HttpMethods[method];
    }
  }
}
