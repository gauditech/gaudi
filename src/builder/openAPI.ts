import { flatMap, mapValues } from "lodash";
import { OpenAPIV3 } from "openapi-types";

import { buildEndpointPath } from "./query";

import { getRef } from "@src/common/refs";
import {
  Definition,
  EndpointDef,
  EntrypointDef,
  FieldsetDef,
  SelectItem,
} from "@src/types/definition";

export function buildOpenAPI(definition: Definition, pathPrefix: string): OpenAPIV3.Document {
  const endpoints = definition.entrypoints.map(extractEndpoints).flat();

  function buildSchema(select: SelectItem[]) {
    const schemaEntries = select.map((select): [string, OpenAPIV3.SchemaObject] => {
      switch (select.kind) {
        case "field": {
          const field = getRef<"field">(definition, select.refKey);
          return [select.alias, { type: convertToOpenAPIType(field.value.type) }];
        }
        // NOTE: not yet implemented
        // case "constant":
        //   return [select.alias, { type: convertToOpenAPIType(select.type) }];
        case "reference":
        case "relation":
        case "query": {
          const isObject = select.kind === "reference";
          const properties = buildSchema(select.select);
          if (isObject) {
            return [select.alias, { type: "object", properties }];
          } else {
            return [select.alias, { type: "array", items: { type: "object", properties } }];
          }
        }
      }
    });

    return Object.fromEntries(schemaEntries);
  }

  function buildEndpointOperation(
    endpoint: EndpointDef,
    hasParameters: boolean
  ): OpenAPIV3.OperationObject {
    const properties = buildSchema(endpoint.response ?? []);
    const isArray = endpoint.kind === "list";

    const objectSchema: OpenAPIV3.SchemaObject = { type: "object", properties };
    const schema: OpenAPIV3.SchemaObject = isArray
      ? { type: "array", items: objectSchema }
      : objectSchema;

    const operation: OpenAPIV3.OperationObject = {
      responses: {
        200: {
          description: "Successful response",
          content: { "application/json": { schema } },
        },
      },
    };

    if (hasParameters) {
      operation.responses[404] = {
        description: "Resource not found",
        content: {
          "application/json": {
            schema: { type: "object", properties: { message: { type: "string" } } },
          },
        },
      };
    }

    if (endpoint.kind === "create") {
      const schema = buildSchemaFromFieldset(endpoint.fieldset);
      operation.requestBody = { content: { "application/json": { schema } } };
    }

    return operation;
  }

  const paths = endpoints.reduce((paths, endpoint) => {
    const gaudiPath = buildEndpointPath(endpoint);
    const method = buildEndpointMethod(endpoint);
    const parameters = gaudiPath.params.map(
      ({ name, type }): OpenAPIV3.ParameterObject => ({
        in: "path",
        name,
        required: true,
        schema: { type: convertToOpenAPIType(type) },
      })
    );
    // FIXME provide path components in definition
    const pathRegex = /:([\w]*)/g;
    const path = pathPrefix + gaudiPath.path.replace(pathRegex, "{$1}");

    const pathItem = paths[path] ?? { parameters };
    pathItem[method] = buildEndpointOperation(endpoint, parameters.length > 0);
    paths[path] = pathItem;

    return paths;
  }, {} as OpenAPIV3.PathsObject);

  return { openapi: "3.0.3", info: { title: "Title", version: "1.0.0" }, paths };
}

function buildSchemaFromFieldset(fieldset: FieldsetDef): OpenAPIV3.SchemaObject {
  if (fieldset.kind === "field") {
    return { type: convertToOpenAPIType(fieldset.type) };
  }

  const schema: OpenAPIV3.SchemaObject = { type: "object" };

  schema.properties = mapValues(fieldset.record, buildSchemaFromFieldset);

  const required = flatMap(fieldset.record, (fieldset, name) => (!fieldset.nullable ? [name] : []));
  if (required.length > 0) {
    schema.required = required;
  }

  return schema;
}

function convertToOpenAPIType(
  type: "boolean" | "integer" | "text"
): OpenAPIV3.NonArraySchemaObjectType {
  switch (type) {
    case "boolean":
    case "integer":
      return type;
    case "text":
      return "string";
  }
}

function extractEndpoints(entrypoint: EntrypointDef): EndpointDef[] {
  const nestedEndpoints = entrypoint.entrypoints.reduce((agg, entrypoint) => {
    return agg.concat(entrypoint.endpoints, extractEndpoints(entrypoint));
  }, [] as EndpointDef[]);
  return [...entrypoint.endpoints, ...nestedEndpoints];
}

function buildEndpointMethod(endpoint: EndpointDef): OpenAPIV3.HttpMethods {
  switch (endpoint.kind) {
    case "list":
      return OpenAPIV3.HttpMethods.GET;
    case "get":
      return OpenAPIV3.HttpMethods.GET;
    case "create":
      return OpenAPIV3.HttpMethods.POST;
    case "update":
      return OpenAPIV3.HttpMethods.PATCH;
    case "delete":
      return OpenAPIV3.HttpMethods.DELETE;
  }
}
