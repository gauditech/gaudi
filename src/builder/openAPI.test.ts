import { buildOpenAPI } from "./openAPI";

import { compileToOldSpec, compose } from "@src/index";

describe("openAPI", () => {
  it("build spec", () => {
    const bp = `
    model Org {
      field slug { type string, unique }
      relation repos { from Repo, through org }
    }
    model Repo {
      reference org { to Org }
      field name { type string }
    }
    auth { method basic {} }
    api Test {
      entrypoint Org as org {
        identify { through slug }

        get endpoint {
          authorize { @auth.id is not null }
        }
        list endpoint { pageable }
        create endpoint {}
        update endpoint {
          authorize { org.id > 4 }
        }
        delete endpoint {
          authorize { org.id > 4 and @auth.id is not null }
        }

        entrypoint repos {
          response { id, name }

          get endpoint {}
          list endpoint { pageable }
          create endpoint {}
          update endpoint {}
          delete endpoint {}
        }
      }
    }
    `;

    const def = compose(compileToOldSpec(bp));

    expect(buildOpenAPI(def)).toMatchSnapshot();
  });
});
