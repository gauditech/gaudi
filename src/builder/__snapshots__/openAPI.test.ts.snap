// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`openAPI build spec 1`] = `
{
  "components": {
    "securitySchemes": {
      "bearerAuth": {
        "scheme": "bearer",
        "type": "http",
      },
    },
  },
  "info": {
    "title": "Title",
    "version": "1.0.0",
  },
  "openapi": "3.0.3",
  "paths": {
    "/api/auth/auth_user/login": {
      "post": {
        "parameters": [],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "properties": {
                  "password": {
                    "type": "string",
                  },
                  "username": {
                    "type": "string",
                  },
                },
                "required": [
                  "username",
                  "password",
                ],
                "type": "object",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {},
                  "required": [],
                  "type": "object",
                },
              },
            },
            "description": "Successful response",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "code": {
                      "enum": [
                        "ERROR_CODE_VALIDATION",
                      ],
                      "type": "string",
                    },
                    "data": {
                      "type": "object",
                    },
                    "message": {
                      "enum": [
                        "Validation error",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Validation error",
          },
        },
        "tags": [
          "Auth",
        ],
      },
    },
    "/api/auth/auth_user/logout": {
      "post": {
        "parameters": [],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "properties": {},
                "type": "object",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {},
                  "required": [],
                  "type": "object",
                },
              },
            },
            "description": "Successful response",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "code": {
                      "enum": [
                        "ERROR_CODE_VALIDATION",
                      ],
                      "type": "string",
                    },
                    "data": {
                      "type": "object",
                    },
                    "message": {
                      "enum": [
                        "Validation error",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Validation error",
          },
        },
        "tags": [
          "Auth",
        ],
      },
    },
    "/api/auth/auth_user/register": {
      "post": {
        "parameters": [],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "properties": {
                  "authUser": {
                    "properties": {
                      "name": {
                        "type": "string",
                      },
                      "username": {
                        "type": "string",
                      },
                    },
                    "required": [
                      "name",
                      "username",
                    ],
                    "type": "object",
                  },
                  "password": {
                    "type": "string",
                  },
                },
                "required": [
                  "authUser",
                  "password",
                ],
                "type": "object",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {},
                  "required": [],
                  "type": "object",
                },
              },
            },
            "description": "Successful response",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "code": {
                      "enum": [
                        "ERROR_CODE_VALIDATION",
                      ],
                      "type": "string",
                    },
                    "data": {
                      "type": "object",
                    },
                    "message": {
                      "enum": [
                        "Validation error",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Validation error",
          },
        },
        "tags": [
          "Auth",
        ],
      },
    },
    "/api/test/org": {
      "get": {
        "parameters": [
          {
            "in": "query",
            "name": "pageSize",
            "required": false,
            "schema": {
              "type": "integer",
            },
          },
          {
            "in": "query",
            "name": "page",
            "required": false,
            "schema": {
              "type": "integer",
            },
          },
        ],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "data": {
                      "items": {
                        "properties": {
                          "id": {
                            "nullable": false,
                            "type": "integer",
                          },
                          "slug": {
                            "nullable": false,
                            "type": "string",
                          },
                        },
                        "required": [
                          "id",
                          "slug",
                        ],
                        "type": "object",
                      },
                      "type": "array",
                    },
                    "page": {
                      "type": "integer",
                    },
                    "pageSize": {
                      "type": "integer",
                    },
                    "totalCount": {
                      "type": "integer",
                    },
                    "totalPages": {
                      "type": "integer",
                    },
                  },
                  "required": [
                    "page",
                    "pageSize",
                    "totalPages",
                    "totalCount",
                    "data",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Successful response",
          },
        },
        "tags": [
          "Test",
        ],
      },
      "post": {
        "parameters": [],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "properties": {
                  "slug": {
                    "type": "string",
                  },
                },
                "required": [
                  "slug",
                ],
                "type": "object",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "id": {
                      "nullable": false,
                      "type": "integer",
                    },
                    "slug": {
                      "nullable": false,
                      "type": "string",
                    },
                  },
                  "required": [
                    "id",
                    "slug",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Successful response",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "code": {
                      "enum": [
                        "ERROR_CODE_VALIDATION",
                      ],
                      "type": "string",
                    },
                    "data": {
                      "type": "object",
                    },
                    "message": {
                      "enum": [
                        "Validation error",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Validation error",
          },
        },
        "tags": [
          "Test",
        ],
      },
    },
    "/api/test/org/{org_slug}": {
      "delete": {
        "parameters": [
          {
            "in": "path",
            "name": "org_slug",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "204": {
            "description": "Successful response",
          },
          "401": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "message": {
                      "enum": [
                        "ERROR_CODE_UNAUTHENTICATED",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Unauthenticated",
          },
          "403": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "message": {
                      "enum": [
                        "ERROR_CODE_FORBIDDEN",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Unauthorized",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "message": {
                      "enum": [
                        "ERROR_CODE_RESOURCE_NOT_FOUND",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Resource not found",
          },
        },
        "security": [
          {
            "bearerAuth": [],
          },
        ],
        "tags": [
          "Test",
        ],
      },
      "get": {
        "parameters": [
          {
            "in": "path",
            "name": "org_slug",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "id": {
                      "nullable": false,
                      "type": "integer",
                    },
                    "slug": {
                      "nullable": false,
                      "type": "string",
                    },
                  },
                  "required": [
                    "id",
                    "slug",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Successful response",
          },
          "401": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "message": {
                      "enum": [
                        "ERROR_CODE_UNAUTHENTICATED",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Unauthenticated",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "message": {
                      "enum": [
                        "ERROR_CODE_RESOURCE_NOT_FOUND",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Resource not found",
          },
        },
        "security": [
          {
            "bearerAuth": [],
          },
        ],
        "tags": [
          "Test",
        ],
      },
      "patch": {
        "parameters": [
          {
            "in": "path",
            "name": "org_slug",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "properties": {
                  "slug": {
                    "type": "string",
                  },
                },
                "required": [
                  "slug",
                ],
                "type": "object",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "id": {
                      "nullable": false,
                      "type": "integer",
                    },
                    "slug": {
                      "nullable": false,
                      "type": "string",
                    },
                  },
                  "required": [
                    "id",
                    "slug",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Successful response",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "code": {
                      "enum": [
                        "ERROR_CODE_VALIDATION",
                      ],
                      "type": "string",
                    },
                    "data": {
                      "type": "object",
                    },
                    "message": {
                      "enum": [
                        "Validation error",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Validation error",
          },
          "403": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "message": {
                      "enum": [
                        "ERROR_CODE_FORBIDDEN",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Unauthorized",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "message": {
                      "enum": [
                        "ERROR_CODE_RESOURCE_NOT_FOUND",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Resource not found",
          },
        },
        "security": [
          {
            "bearerAuth": [],
          },
        ],
        "tags": [
          "Test",
        ],
      },
    },
    "/api/test/org/{org_slug}/repos": {
      "get": {
        "parameters": [
          {
            "in": "path",
            "name": "org_slug",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
          {
            "in": "query",
            "name": "pageSize",
            "required": false,
            "schema": {
              "type": "integer",
            },
          },
          {
            "in": "query",
            "name": "page",
            "required": false,
            "schema": {
              "type": "integer",
            },
          },
        ],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "data": {
                      "items": {
                        "properties": {
                          "id": {
                            "nullable": false,
                            "type": "integer",
                          },
                          "name": {
                            "nullable": false,
                            "type": "string",
                          },
                        },
                        "required": [
                          "id",
                          "name",
                        ],
                        "type": "object",
                      },
                      "type": "array",
                    },
                    "page": {
                      "type": "integer",
                    },
                    "pageSize": {
                      "type": "integer",
                    },
                    "totalCount": {
                      "type": "integer",
                    },
                    "totalPages": {
                      "type": "integer",
                    },
                  },
                  "required": [
                    "page",
                    "pageSize",
                    "totalPages",
                    "totalCount",
                    "data",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Successful response",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "message": {
                      "enum": [
                        "ERROR_CODE_RESOURCE_NOT_FOUND",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Resource not found",
          },
        },
        "tags": [
          "Test",
        ],
      },
      "post": {
        "parameters": [
          {
            "in": "path",
            "name": "org_slug",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "properties": {
                  "name": {
                    "type": "string",
                  },
                },
                "required": [
                  "name",
                ],
                "type": "object",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "id": {
                      "nullable": false,
                      "type": "integer",
                    },
                    "name": {
                      "nullable": false,
                      "type": "string",
                    },
                  },
                  "required": [
                    "id",
                    "name",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Successful response",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "code": {
                      "enum": [
                        "ERROR_CODE_VALIDATION",
                      ],
                      "type": "string",
                    },
                    "data": {
                      "type": "object",
                    },
                    "message": {
                      "enum": [
                        "Validation error",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Validation error",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "message": {
                      "enum": [
                        "ERROR_CODE_RESOURCE_NOT_FOUND",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Resource not found",
          },
        },
        "tags": [
          "Test",
        ],
      },
    },
    "/api/test/org/{org_slug}/repos/{repo_id}": {
      "delete": {
        "parameters": [
          {
            "in": "path",
            "name": "org_slug",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
          {
            "in": "path",
            "name": "repo_id",
            "required": true,
            "schema": {
              "type": "integer",
            },
          },
        ],
        "responses": {
          "204": {
            "description": "Successful response",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "message": {
                      "enum": [
                        "ERROR_CODE_RESOURCE_NOT_FOUND",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Resource not found",
          },
        },
        "tags": [
          "Test",
        ],
      },
      "get": {
        "parameters": [
          {
            "in": "path",
            "name": "org_slug",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
          {
            "in": "path",
            "name": "repo_id",
            "required": true,
            "schema": {
              "type": "integer",
            },
          },
        ],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "id": {
                      "nullable": false,
                      "type": "integer",
                    },
                    "name": {
                      "nullable": false,
                      "type": "string",
                    },
                  },
                  "required": [
                    "id",
                    "name",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Successful response",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "message": {
                      "enum": [
                        "ERROR_CODE_RESOURCE_NOT_FOUND",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Resource not found",
          },
        },
        "tags": [
          "Test",
        ],
      },
      "patch": {
        "parameters": [
          {
            "in": "path",
            "name": "org_slug",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
          {
            "in": "path",
            "name": "repo_id",
            "required": true,
            "schema": {
              "type": "integer",
            },
          },
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "properties": {
                  "name": {
                    "type": "string",
                  },
                  "org_id": {
                    "type": "integer",
                  },
                },
                "required": [
                  "name",
                  "org_id",
                ],
                "type": "object",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "id": {
                      "nullable": false,
                      "type": "integer",
                    },
                    "name": {
                      "nullable": false,
                      "type": "string",
                    },
                  },
                  "required": [
                    "id",
                    "name",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Successful response",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "code": {
                      "enum": [
                        "ERROR_CODE_VALIDATION",
                      ],
                      "type": "string",
                    },
                    "data": {
                      "type": "object",
                    },
                    "message": {
                      "enum": [
                        "Validation error",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Validation error",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "message": {
                      "enum": [
                        "ERROR_CODE_RESOURCE_NOT_FOUND",
                      ],
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Resource not found",
          },
        },
        "tags": [
          "Test",
        ],
      },
    },
  },
}
`;
